src/discord_bot/bot.cpp

```diff
@@ -3,23 +3,176 @@
 
 #include "eelbot_framework/discord_bot/bot.hpp"
 
+#include "eelbot_framework/compression.hpp"
+#include "eelbot_framework/discord_bot/endpoints.hpp"
 #include "eelbot_framework/discord_bot/rest_api.hpp"
 #include "eelbot_framework/json.hpp"
+#include "eelbot_framework/os.hpp"
+#include "eelbot_framework/websocket.hpp"
 #include "fmt/core.h"
 
 #include <stdexcept>
+#include <thread>
 namespace eelbot_framework {
 
 namespace discord_bot {
 
-http_request_settings bot::get_http_request_settings() {
+http_request_settings bot::get_http_request_settings() noexcept {
 	http_request_settings request_settings;
 	request_settings.tls_ver      = this->tls_ver;
 	request_settings.ca_directory = this->ca_directory;
 	request_settings.proxy        = this->http_proxy;
 	return request_settings;
 }
 
+void bot::send_heartbeat() {
+	payload pl;
+	pl.op = opcode::heartbeat;
+	pl.d  = this->sequence_num;
+	this->ws_client->send(to_json_str(pl));
+}
+
+void bot::send_identify() {
+	// TODO: Make shard info configurable down the line.
+	shard_info sh_info;
+	sh_info.shard_id   = 0;
+	sh_info.num_shards = 1;
+
+	identify_connection_properties id_conn_props;
+	id_conn_props.os      = get_platform_name();
+	id_conn_props.browser = "Eelbot Framework";
+	id_conn_props.device  = "Eelbot Framework";
+
+	identify id;
+	id.token      = this->token;
+	id.intents    = 513; // TODO: Determine this dynamically.
+	id.shard      = sh_info;
+	id.properties = id_conn_props;
+	id.compress   = false;
+
+	this->logger->log(log::level::DEBUG, "Sending identify.");
+	payload pl;
+	pl.op = opcode::identify;
+	pl.d  = id;
+	this->ws_client->send(to_json_str(pl));
+}
+
+void bot::send_resume() {
+	resume rs;
+	rs.token = this->token;
+	rs.seq   = this->sequence_num;
+
+	{
+		std::scoped_lock lock(this->session_mutex);
+		rs.session_id = this->session_id;
+	}
+
+	this->logger->log(log::level::DEBUG, "Sending resume.");
+	payload pl;
+	pl.op = opcode::resume;
+	pl.d  = rs;
+	this->ws_client->send(to_json_str(pl));
+}
+
+void bot::handle_message(const std::string &message) noexcept try {
+	bool flush = false;
+	if (message.size() >= 4) {
+		flush = message[message.size() - 4] == '\x00' && message[message.size() - 3] == '\x00' &&
+		        message[message.size() - 2] == '\xFF' && message[message.size() - 1] == '\xFF';
+	}
+
+	this->z_inflator->push(message);
+	if (!flush) {
+		this->logger->log(log::level::TRACE, "Zlib suffix not received; returning without parsing.");
+		return;
+	}
+
+	payload pl = parse_from_json<payload>(this->z_inflator->get_result());
+
+	// If the gateway sent a sequence number, update the cached sequence value.
+	if (pl.s.has_value()) {
+		this->sequence_num = pl.s.value();
+		this->logger->log(log::level::TRACE, fmt::format("Sequence updated to {}.", this->sequence_num));
+	}
+
+	switch (pl.op) {
```

Comment:
interesting, this lack of indentation for the case statements is intentional?

Reply:
Yes it's intentional. It comes down to personal preference, and I'm sure clang-format supports both methods.